src/main/java/technology/dice/dicefairlink/discovery/members/FairlinkMemberFinder.java

```diff
@@ -0,0 +1,163 @@
+    tagsPollingExecutor.scheduleAtFixedRate(
+        () -> excludedInstanceIds = safeExclusionsDiscovery(),
+        startJitter.getSeconds(),
+        fairlinkConfiguration.getTagsPollerInterval().getSeconds(),
+        TimeUnit.SECONDS);
+  }
+
+  private Set<String> safeExclusionsDiscovery() {
+    try {
+      return tagFilter.listExcludedInstances(EXCLUSION_TAG);
+    } catch (Exception e) {
+      LOGGER.log(
+          Level.SEVERE, "Could not discover exclusions; including all discovered instances", e);
+      return EMPTY_SET;
+    }
+  }
+
+  public final SizedIterator<String> discoverReplicas() {
+    try {
+      long before = System.currentTimeMillis();
+      final ClusterInfo clusterInfo = this.memberFinder.discoverCluster();
+      this.fallbackEndpoint = Optional.of(clusterInfo.getReadonlyEndpoint());
+      final Set<String> filteredReplicas =
+          clusterInfo.getReplicas().stream()
+              .filter(db -> !excludedInstanceIds.contains(db))
+              .filter(
+                  dbIdentifier ->
+                      (!this.fairlinkConfiguration.isValidateConnection())
+                          || this.validate(fairlinkConfiguration.hostname(dbIdentifier)))
+              .map(dbIdentifier -> fairlinkConfiguration.hostname(dbIdentifier))
+              .collect(Collectors.toSet());
+      final SizedIterator<String> result =
+          filteredReplicas.isEmpty()
+              ? this.iteratorBuilder.apply(this.setOf(clusterInfo.getReadonlyEndpoint()))
+              : this.iteratorBuilder.apply(filteredReplicas);
+      long after = System.currentTimeMillis();
+      LOGGER.info(
+          "Updated list of replicas in "
+              + (after - before)
+              + " ms. Found "
+              + filteredReplicas.size()
+              + " good, active, non-excluded replica"
+              + (filteredReplicas.size() != 1 ? "s" : "")
+              + " (validation "
+              + (fairlinkConfiguration.isValidateConnection() ? "" : "NOT ")
+              + "done). Excluded "
+              + this.excludedInstanceIds.size()
+              + " instance"
+              + (this.excludedInstanceIds.size() != 1 ? "" : "s")
+              + ". Next update in "
+              + this.fairlinkConfiguration.getReplicaPollInterval());
+      return result;
+    } catch (Exception e) {
+      LOGGER.log(
+          Level.WARNING,
+          "Error discovering cluster identified by ["
+              + this.fairlinkConnectionString.getFairlinkUri()
+              + "]. Will return fallback endpoint"
+              + this.fallbackEndpoint
+              + " if available",
+          e);
+      if (!this.fallbackEndpoint.isPresent()) {
+        LOGGER.log(
+            Level.SEVERE,
+            "Fallback endpoint not available. This means the cluster has never been successfully discovered. This is probably a permanent error condition");
```

Comment:
Throw after this some sort of Exception?

Reply:
It was doing it on the following statement. 
But I re-wrote it in a slightly more clear way